### PR TITLE
Use IDXGIFactory6_EnumAdapterByGpuPreference to find the High Performance gpu

### PR DESCRIPTION
**Pull request by bartwe**
Use EnumAdapterByGpuPreference 

https://docs.microsoft.com/en-us/windows/win32/api/dxgi1_6/nf-dxgi1_6-idxgifactory6-enumadapterbygpupreference

(note include path reordering)

Implemented such that it will fallback to the current implementation in version of windows before Windows 10, version 1803 

https://github.com/libsdl-org/SDL/issues/5210

### CODE REVIEW

**Comment by flibitijibibo** [comment]
Similar to the D3D 11.1 features we use we'll need to manually declare the DXGI 1.6 features, since this breaks when building with the Jun2010 DXSDK. Other than that this seems like a good idea, so let's get this in.

**Comment by bartwe** [comment]
Is there an example on how to do this for the interface/vtable ? the other bits are already moved into local declarations

**Comment by flibitijibibo** [comment]
Our declaration of user annotations:

https://github.com/FNA-XNA/FNA3D/blob/master/src/FNA3D_Driver_D3D11.h#L68

**Comment by bartwe** [comment]
Ok changes made and it is now no longer dependent on a new sdk.

File: src/FNA3D_Driver_D3D11.c [hunk 1]
```diff
@@ -5352,11 +5352,31 @@ static void D3D11_PLATFORM_GetDefaultAdapter(
...
 ) {
-	IDXGIFactory1_EnumAdapters1(
+	void* factory6;
+	HRESULT res;
+	res = IDXGIFactory1_QueryInterface(
 		(IDXGIFactory1*) factory,
-		0,
-		adapter
+		&D3D_IID_IDXGIFactory6,
+		(void**)&factory6
```
Comment (flibitijibibo) [inline src/FNA3D_Driver_D3D11.c:5360]:
```suggestion
		(void**) &factory6
```

File: src/FNA3D_Driver_D3D11.h [hunk 2]
```diff
@@ -123,4 +125,191 @@ struct ID3DUserDefinedAnnotation
...
+        DXGI_FEATURE_PRESENT_ALLOW_TEARING	= 0
+    } 	DXGI_FEATURE;
+
+typedef 
+enum DXGI_GPU_PREFERENCE
+    {
+        DXGI_GPU_PREFERENCE_UNSPECIFIED	= 0,
+        DXGI_GPU_PREFERENCE_MINIMUM_POWER	= ( DXGI_GPU_PREFERENCE_UNSPECIFIED + 1 ) ,
+        DXGI_GPU_PREFERENCE_HIGH_PERFORMANCE	= ( DXGI_GPU_PREFERENCE_MINIMUM_POWER + 1 ) 
+    } 	DXGI_GPU_PREFERENCE;
```
Comment (flibitijibibo) [inline src/FNA3D_Driver_D3D11.h:144]:
```suggestion
/* From igdx1_6.h, cleaned up a bit... */

typedef enum
{
	DXGI_FEATURE_PRESENT_ALLOW_TEARING	= 0
} DXGI_FEATURE;

typedef enum
{
	DXGI_GPU_PREFERENCE_UNSPECIFIED	= 0,
	DXGI_GPU_PREFERENCE_MINIMUM_POWER	= ( DXGI_GPU_PREFERENCE_UNSPECIFIED + 1 ) ,
	DXGI_GPU_PREFERENCE_HIGH_PERFORMANCE	= ( DXGI_GPU_PREFERENCE_MINIMUM_POWER + 1 ) 
} DXGI_GPU_PREFERENCE;
```

File: src/FNA3D_Driver_D3D11.h [hunk 3]
```diff
@@ -123,4 +125,191 @@ struct ID3DUserDefinedAnnotation
...
+
+} IDXGIFactory6Vtbl;
+
+struct IDXGIFactory6
+{
+	struct IDXGIFactory6Vtbl *lpVtbl;
+};
+
+#define IDXGIFactory6_EnumAdapterByGpuPreference(This,Adapter,GpuPreference,riid,ppvAdapter)	\
+    ( (This)->lpVtbl -> EnumAdapterByGpuPreference(This,Adapter,GpuPreference,riid,ppvAdapter) )
```
Comment (flibitijibibo) [inline src/FNA3D_Driver_D3D11.h:312]:
```suggestion
	( (This)->lpVtbl -> EnumAdapterByGpuPreference(This,Adapter,GpuPreference,riid,ppvAdapter) ) 
```

File: src/FNA3D_Driver_D3D11.h [hunk 4]
```diff
@@ -123,4 +125,191 @@ struct ID3DUserDefinedAnnotation
...
+
+struct IDXGIFactory6
+{
+	struct IDXGIFactory6Vtbl *lpVtbl;
+};
+
+#define IDXGIFactory6_EnumAdapterByGpuPreference(This,Adapter,GpuPreference,riid,ppvAdapter)	\
+    ( (This)->lpVtbl -> EnumAdapterByGpuPreference(This,Adapter,GpuPreference,riid,ppvAdapter) ) 
+
+
```
Comment (flibitijibibo) [inline src/FNA3D_Driver_D3D11.h:314]:
```suggestion
```

**Review by flibitijibibo — COMMENTED** [review]
Just a bunch of whitespace... once this is applied I'll merge this in!

File: visualc/FNA3D.vcxproj [hunk 5]
```diff
@@ -42,7 +42,7 @@
   </ImportGroup>
   <PropertyGroup Label="UserMacros" />
   <PropertyGroup>
-    <IncludePath>..\..\SDL2\include;..\MojoShader;..\include;$(DXSDK_DIR)\Include;..\Vulkan-Headers\include;$(IncludePath)</IncludePath>
+    <IncludePath>..\..\SDL2\include;..\MojoShader;..\include;..\Vulkan-Headers\include;$(IncludePath);$(DXSDK_DIR)\Include</IncludePath>
```
Comment (flibitijibibo) [inline visualc/FNA3D.vcxproj:45]: Woops, missed this... should probably undo this.

File: src/FNA3D_Driver_D3D11.h [hunk 6]
```diff
@@ -123,4 +125,187 @@ struct ID3DUserDefinedAnnotation
...
+        UINT FeatureSupportDataSize);
+        
+    HRESULT ( STDMETHODCALLTYPE *EnumAdapterByGpuPreference )( 
+        IDXGIFactory6 * This,
+        UINT Adapter,
+        DXGI_GPU_PREFERENCE GpuPreference,
+        REFIID riid,
+        void **ppvAdapter);
+
+} IDXGIFactory6Vtbl;
```
Comment (flibitijibibo) [inline src/FNA3D_Driver_D3D11.h:301]:
```suggestion
	HRESULT ( STDMETHODCALLTYPE *QueryInterface )( 
		IDXGIFactory6 * This,
		REFIID riid,
		void **ppvObject);
		
	ULONG ( STDMETHODCALLTYPE *AddRef )( 
		IDXGIFactory6 * This);
		
	ULONG ( STDMETHODCALLTYPE *Release )( 
		IDXGIFactory6 * This);
		
	HRESULT ( STDMETHODCALLTYPE *SetPrivateData )( 
		IDXGIFactory6 * This,
		REFGUID Name,
		UINT DataSize,
		const void *pData);
		
	HRESULT ( STDMETHODCALLTYPE *SetPrivateDataInterface )( 
		IDXGIFactory6 * This,
		REFGUID Name,
		const IUnknown *pUnknown);
		
	HRESULT ( STDMETHODCALLTYPE *GetPrivateData )( 
		IDXGIFactory6 * This,
		REFGUID Name,
		UINT *pDataSize,
		void *pData);
		
	HRESULT ( STDMETHODCALLTYPE *GetParent )( 
		IDXGIFactory6 * This,
		REFIID riid,
		void **ppParent);
		
	HRESULT ( STDMETHODCALLTYPE *EnumAdapters )( 
		IDXGIFactory6 * This,
		UINT Adapter,
		IDXGIAdapter **ppAdapter);
		
	HRESULT ( STDMETHODCALLTYPE *MakeWindowAssociation )( 
		IDXGIFactory6 * This,
		HWND WindowHandle,
		UINT Flags);
		
	HRESULT ( STDMETHODCALLTYPE *GetWindowAssociation )( 
		IDXGIFactory6 * This,
		HWND *pWindowHandle);
		
	HRESULT ( STDMETHODCALLTYPE *CreateSwapChain )( 
		IDXGIFactory6 * This,
		IUnknown *pDevice,
		DXGI_SWAP_CHAIN_DESC *pDesc,
		IDXGISwapChain **ppSwapChain);
		
	HRESULT ( STDMETHODCALLTYPE *CreateSoftwareAdapter )( 
		IDXGIFactory6 * This,
		HMODULE Module,
		IDXGIAdapter **ppAdapter);
		
	HRESULT ( STDMETHODCALLTYPE *EnumAdapters1 )( 
		IDXGIFactory6 * This,
		UINT Adapter,
		IDXGIAdapter1 **ppAdapter);
		
	BOOL ( STDMETHODCALLTYPE *IsCurrent )( 
		IDXGIFactory6 * This);
		
	BOOL ( STDMETHODCALLTYPE *IsWindowedStereoEnabled )( 
		IDXGIFactory6 * This);
		
	HRESULT ( STDMETHODCALLTYPE *CreateSwapChainForHwnd )( 
		IDXGIFactory6 * This,
		IUnknown *pDevice,
		HWND hWnd,
		void *pDesc,
		void *pFullscreenDesc,
		void *pRestrictToOutput,
		void **ppSwapChain);
		
	HRESULT ( STDMETHODCALLTYPE *CreateSwapChainForCoreWindow )( 
		IDXGIFactory6 * This,
		IUnknown *pDevice,
		IUnknown *pWindow,
		void *pDesc,
		void *pRestrictToOutput,
		void **ppSwapChain);
		
	HRESULT ( STDMETHODCALLTYPE *GetSharedResourceAdapterLuid )( 
		IDXGIFactory6 * This,
		HANDLE hResource,
		LUID *pLuid);
		
	HRESULT ( STDMETHODCALLTYPE *RegisterStereoStatusWindow )( 
		IDXGIFactory6 * This,
		HWND WindowHandle,
		UINT wMsg,
		DWORD *pdwCookie);
		
	HRESULT ( STDMETHODCALLTYPE *RegisterStereoStatusEvent )( 
		IDXGIFactory6 * This,
		HANDLE hEvent,
		DWORD *pdwCookie);
		
	void ( STDMETHODCALLTYPE *UnregisterStereoStatus )( 
		IDXGIFactory6 * This,
		DWORD dwCookie);
		
	HRESULT ( STDMETHODCALLTYPE *RegisterOcclusionStatusWindow )( 
		IDXGIFactory6 * This,
		HWND WindowHandle,
		UINT wMsg,
		DWORD *pdwCookie);
		
	HRESULT ( STDMETHODCALLTYPE *RegisterOcclusionStatusEvent )( 
		IDXGIFactory6 * This,
		HANDLE hEvent,
		DWORD *pdwCookie);
		
	void ( STDMETHODCALLTYPE *UnregisterOcclusionStatus )( 
		IDXGIFactory6 * This,
		DWORD dwCookie);
		
	HRESULT ( STDMETHODCALLTYPE *CreateSwapChainForComposition )( 
		IDXGIFactory6 * This,
		IUnknown *pDevice,
		void *pDesc,
		void *pRestrictToOutput,
		void **ppSwapChain);
		
	UINT ( STDMETHODCALLTYPE *GetCreationFlags )( 
		IDXGIFactory6 * This);
		
	HRESULT ( STDMETHODCALLTYPE *EnumAdapterByLuid )( 
		IDXGIFactory6 * This,
		LUID AdapterLuid,
		REFIID riid,
		void **ppvAdapter);
		
	HRESULT ( STDMETHODCALLTYPE *EnumWarpAdapter )( 
		IDXGIFactory6 * This,
		REFIID riid,
		void **ppvAdapter);
		
	HRESULT ( STDMETHODCALLTYPE *CheckFeatureSupport )( 
		IDXGIFactory6 * This,
		DXGI_FEATURE Feature,
		void *pFeatureSupportData,
		UINT FeatureSupportDataSize);
		
	HRESULT ( STDMETHODCALLTYPE *EnumAdapterByGpuPreference )( 
		IDXGIFactory6 * This,
		UINT Adapter,
		DXGI_GPU_PREFERENCE GpuPreference,
		REFIID riid,
		void **ppvAdapter);
} IDXGIFactory6Vtbl;
```

**Pull request merged**